File: cpp/minsk/source/analysis/diagnostic_bag.cpp
#include "minsk/analysis/diagnostic_bag.hpp"
#include "fmt/format.h"
#include "magic_enum.hpp"
#include "minsk/analysis/text/span.hpp"

using minsk::analysis::diagnostic;
using minsk::analysis::diagnostic_bag;
using minsk::analysis::text::text_span;
using minsk::runtime::object_kind;

void diagnostic_bag::report(text_span span, std::string &&message) {
  m_diagnostics.emplace_back(span, std::move(message));
}

diagnostic_bag::iterator diagnostic_bag::begin() const {
  return m_diagnostics.cbegin();
}

diagnostic_bag::iterator diagnostic_bag::end() const {
  return m_diagnostics.end();
}

void diagnostic_bag::push_back(const value_type &v) {
  m_diagnostics.push_back(v);
}

const diagnostic &diagnostic_bag::operator[](size_type i) const {
  return m_diagnostics[i];
}

bool diagnostic_bag::empty() const { return m_diagnostics.empty(); }

diagnostic_bag::size_type diagnostic_bag::size() const {
  return m_diagnostics.size();
}

void diagnostic_bag::report_invalid_int(text_span span, std::string_view text) {
  report(span, fmt::format("The number '{}' is not a valid int", text));
}

void diagnostic_bag::report_bad_character(int position, char character) {
  report(text_span{position, 1},
         fmt::format("Bad character in input: '{}'", character));
}

void diagnostic_bag::report_unexpected_token(text_span span,
                                             syntax::syntax_kind expected_kind,
                                             syntax::syntax_kind actual_kind) {
  report(span, fmt::format("Expected next token to be <{}>, got <{}> instead",
                           magic_enum::enum_name(expected_kind),
                           magic_enum::enum_name(actual_kind)));
}

void diagnostic_bag::report_undefined_binary_operator(
    text_span span, std::string_view operator_text, object_kind left_type,
    object_kind right_type) {
  report(span, fmt::format(
                   "Binary operator '{}' isn't defined for types '{}' and '{}'",
                   operator_text, magic_enum::enum_name(left_type),
                   magic_enum::enum_name(right_type)));
}

void diagnostic_bag::report_undefined_unary_operator(
    text_span span, std::string_view operator_text, object_kind operand_type) {
  report(span, fmt::format("Unary operator '{}' isn't defined for type '{}'",
                           operator_text, magic_enum::enum_name(operand_type)));
}

void diagnostic_bag::report_undefined_name(text_span span,
                                           std::string_view name) {
  report(span, fmt::format("Name '{}' is undeclared", name));
}

void diagnostic_bag::report_variable_already_declared(text_span span,
                                                      std::string_view name) {
  report(span, fmt::format("Name '{}' is already declared", name));
}

void diagnostic_bag::report_cannot_convert(text_span span,
                                           object_kind from_type,
                                           object_kind to_type) {
  report(span, fmt::format("Cannot convert type '{}' to '{}'",
                           magic_enum::enum_name(from_type),
                           magic_enum::enum_name(to_type)));
}

void diagnostic_bag::report_cannot_assign(text_span span,
                                          std::string_view name) {
  report(span, fmt::format("Variable '{}' is read-only", name));
}

File: ocaml/lib/code_analysis/diagnostic_bag.ml
open Runtime

let sprintf = Printf.sprintf

type t = { mutable diagnostics : Diagnostic.t BatVect.t }

let make () = { diagnostics = BatVect.empty }
let add_range x y = x.diagnostics <- BatVect.concat x.diagnostics y.diagnostics
let to_array x = BatVect.to_array x.diagnostics

let report span message x =
  x.diagnostics <- BatVect.append (Diagnostic.make span message) x.diagnostics

let report_invalid_number span ~text ~ty =
  let message =
    sprintf "The number %s isn't a valid %s." text (Value.show_ty ty)
  in
  report span message

let report_bad_character ~position ~character =
  let span = Text.Span.make position 1 in
  let message = sprintf "Bad character in input: '%c'" character in
  report span message

let report_unexpected_token span ~actual_kind ~expected_kind =
  let message =
    sprintf "Unexpected token <%s>, expected <%s>."
      (Token.show_kind actual_kind)
      (Token.show_kind expected_kind)
  in
  report span message

let report_undefined_unary_operator span ~operator_text ~operand_ty =
  let message =
    sprintf "The unary operator '%s' is not defined for type %s." operator_text
      (Value.show_ty operand_ty)
  in
  report span message

let report_undefined_binary_operator span ~operator_text ~left_ty ~right_ty =
  let message =
    sprintf "The binary operator '%s' is not defined for types %s and %s."
      operator_text (Value.show_ty left_ty) (Value.show_ty right_ty)
  in
  report span message

let report_undefined_name span ~name =
  let message = sprintf "Variable '%s' doesn't exist." name in
  report span message
